src/main/java/org/candlepin/model/Pool.java

```diff
@@ -838,6 +841,10 @@ public Pool setActiveSubscription(Boolean activeSubscription) {
      *  otherwise
      */
     public String getProductUuid() {
+        if (this.productUuid == null || Hibernate.isInitialized(this.product)) {
```

Comment:
Few things here:
- While we do have the UUID available as a field, the deferment to product makes this more of a utility method. Fetching of the product should be done through `.getProduct()` not through a direct field reference.
- Hibernate.isInitialized will return `true` on a null field, and the usage of `this.product` into `getProduct()` could lead to a NPE or otherwise inconsistent output.
- Minorist of nano-optimizations: given that our primary concern is the presence and initialization of product, that should be the first thing we check; the product_uuid field is now our secondary. Granted, I'd expect a JIT compiler to notice this and rearrange stuff as appropriate, it's still good habit to be aware of that kind of thing, and it makes the code a bit more obvious in where the priorities lie.

Given that we need to be concerned with the nullness of our product field, the checks should look something like this:

```
Product product = this.getProduct();

// If we don't have a product instance, it doesn't matter what the 
// productUuid field contains -- we've desync'd and it's likely wrong.
if (product == null) {
  return null;
}

// Return the productUuid if, and only if, our product instance is an
// uninitialized Hibernate proxy and we have a non-null UUID value.
// Otherwise, defer to the product itself.
return !Hibernate.isInitialized(product) && this.productUuid != null ?
  this.productUuid :
  product.getUuid();
```

Reply:
This makes sense! I also added those comments as they will be very helpful for others looking at this code.